File: Leetcode/2558 Take Gifts From the Richest Pile.py
import heapq
import math


class Solution:
    def pickGifts(self, gifts, k):
        max_heap = [-gift for gift in gifts]
        heapq.heapify(max_heap)
        for _ in range(k):
            gift = -heapq.heappop(max_heap)
            gift = -int(math.sqrt(gift))
            heapq.heappush(max_heap, gift)
        return -sum(max_heap)

File: Leetcode/2558 Take Gifts From the Richest Pile.java
class Solution {
    public long pickGifts(int[] gifts, int k) {
        PriorityQueue<Long> pq = new PriorityQueue<>(Comparator.reverseOrder());
        for (int gift : gifts) {
            pq.add((long) gift);
        }
        for (int i = 0; i < k && !pq.isEmpty(); i++) {
            long max = pq.poll();
            pq.add((long) Math.sqrt(max));
        }
        return pq.stream().mapToLong(n -> n).sum();
    }
}
